### DISCUSSION TITLE
Commands like dotnet restore hangs when attempting to emulate `linux/amd64` on a mac

### DISCUSSION THREAD

**Discussion by kurt-mueller-osumc** [dotnet/dotnet-docker#4944] [General]
I created a simple repo here: https://github.com/kurt-mueller-osumc/dotnet-amd64

This repository contains a simple dotnet project and Dockerfile. It's meant to demonstrate that commands like `dotnet restore` & `dotnet build` fail (i.e. they simply hang) when attempting to emulate `linux/amd64` on a `arm64` architecture (i.e. Macs).

Why want to emulate `linux/amd64` on a Mac? In my case, it's because I want to run dotnet isolated Azure Functions on my Mac. Azure Functions are `x64/amd64` only and its official [docker images](https://hub.docker.com/_/microsoft-azure-functions-dotnet-isolated) only support `amd64` architecture.

You can experiencing this by running the following command on a Mac:

```bash
# this process will stall when running `dotnet restore`
docker build --platform linux/amd64 .
```

I'm frustrated that there's no possible way I can run an dotnet-isolated Azure Functions docker image on my Mac.

**Comment by mthalman** [comment]
This won't work because Docker is using QEMU here which is not supported: https://github.com/dotnet/core/blob/main/release-notes/8.0/supported-os.md#qemu. There isn't a workaround available without having a machine that matches the architecture of the target container.

**Reply by kurt-mueller-osumc** [reply]
<img width="1074" alt="Screenshot 2023-10-16 at 11 56 57 AM" src="https://github.com/dotnet/dotnet-docker/assets/66080261/7da3dcd5-29d8-449a-a569-618129ea9eef">

... but I have Rosetta enabled.

**Reply by mthalman** [reply]
@richlander - Does Rosetta have the same issue here?

**Reply by lbussell** [reply]
Can you try running the restore and build of your library under arm64? Similar to how our multi-platform sample Dockerfiles are written: https://github.com/dotnet/dotnet-docker/blob/01184dc44bc6084e891bddd06a22584cefc218a3/samples/dotnetapp/Dockerfile

The Dockerfile in your repro repo would be:
```
FROM --platform=$BUILDPLATFORM mcr.microsoft.com/dotnet/sdk:8.0

ARG TARGETARCH

ENV WORKDIR="/app"

COPY . ${WORKDIR}

WORKDIR  ${WORKDIR}

RUN dotnet restore -a ${TARGETARCH}
```

Then if you build with the same command, it will restore amd64 bits while running on arm64, no emulation. Then you can copy the amd64 restore/publish artifacts to your amd64 (or TARGETARCH) image layer to create your final image. I don't know if you will encounter any other issues running .NET apps under Rosetta though. This could be a NuGet issue.

**Reply by richlander** [reply]
In my testing, Rosetta works better than QEMU. The Rosetta support has been there for a while now.
